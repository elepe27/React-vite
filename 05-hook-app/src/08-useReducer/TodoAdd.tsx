import { useForm } from "../hooks"


export const TodoAdd = ({onNewTodo}:any) => {

    const { formState , onInputChange, onReset, description}:any = useForm( {
        description:''
    });

    const onFormSubmit = ( event:any )=> {

        event.preventDefault();
        
        if(description.length <=1 ) return;
        
        const newTodo = {
            id: new Date().getTime(),
            description,
            done:false,

        };

        onNewTodo(newTodo);
        onReset();
    }

   

    return (
        <form onSubmit={ onFormSubmit }>
            <input
                type='text'
                placeholder="¿Qué hay que hacer?"
                className="form-control"
                name="description"
                value={ description }
                onChange={ onInputChange }
                autoComplete='off'
                
            />

            <button
                type="submit"
                className="btn btn-primary mt-2"
                >
                Agregar
            </button>

        </form>
    )
}
