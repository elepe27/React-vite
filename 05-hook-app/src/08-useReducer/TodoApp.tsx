import { TodoAdd, TodoList  } from "./";
import { useTodo } from '../hooks/useTodo';


export const TodoApp = () => {


   const { todos, onNewTodo, onDelete, onToggleTodo }:any = useTodo();

   const pendientes = todos.filter((todo:any)=> !todo.done);

    return (
        <>
            <h1>TodoApp { todos.length }, <small>Pendientes: { pendientes.length }</small></h1>
            <hr />
            <div className="row">
                <div className="col-7">
                    { /**TodoList */}
                    <TodoList 
                        todo = { todos }
                        onDelete= { onDelete }
                        onToggleTodo={ onToggleTodo }/>
                </div>
                <div className="col-5">
                    <h4>Agregar TODO</h4>
                    <hr />
                    <TodoAdd 
                        onNewTodo = { onNewTodo} />
                </div>
            </div>


        </>
    )
}
