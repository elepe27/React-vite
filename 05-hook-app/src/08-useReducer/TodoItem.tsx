type props = {
    todo: {
        description:string,
        done: boolean,
        id: number
    },
    onDelete: Function,
    onToggleTodo: Function
}

export const TodoItem = ({ todo, onDelete, onToggleTodo }:props) => {
    return (
        <li className="list-group-item d-flex justify-content-between">
            <span 
                
                className={`align-self-center ${ (todo.done) ? 'text-decoration-line-through' :''}`}
                onClick={ () => onToggleTodo( todo.id )}>
                     { todo.description } 
            </span>
            <button  className="btn btn-sm btn-danger" onClick={()=> onDelete(todo.id)}> Borrar </button>
        </li>
    )
}
