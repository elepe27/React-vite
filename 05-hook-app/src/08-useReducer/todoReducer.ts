

export const todoReducer = ( initialState:Array<any> = [], action:any = {} )=> {


    switch ( action.type ) {
        case '[TODO] Add Todo':
                return [  ...initialState, action.payload];
        case '[TODO] Remove Todo':
                return initialState.filter((todo:any) => todo.id !== action.payload );
        case '[TODO] Toggle':
                return initialState.map((todo:any)=> {
                    if (todo.id=== action.payload) {

                        return {
                            ...todo,
                            done: !todo.done
                        }
                    }
                    return todo;
                })


    
        default:
            return initialState;
    }

}