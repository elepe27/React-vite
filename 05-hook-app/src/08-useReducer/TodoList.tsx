import { TodoItem } from "./TodoItem"

type props = {
   todo: Array<any>,
   onDelete: Function,
   onToggleTodo: Function
}

export const TodoList = ( { todo , onDelete, onToggleTodo }:props) => {
    return (
        <ul className="list-group">
            {
                todo.map((value:any)=> <TodoItem key={value.id} todo={ value } onDelete= { onDelete } onToggleTodo= { onToggleTodo }  />)
            }
        </ul>
    )
}
