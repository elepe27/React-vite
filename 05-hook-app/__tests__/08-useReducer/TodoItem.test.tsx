import { render, screen, fireEvent } from '@testing-library/react';
import { TodoItem } from '../../src/08-useReducer/TodoItem';
import * as React from 'react';



describe('Pruebas en <TodoItem/>', () => {

    const todo = {
        id: 1,
        description: 'TODO NUEVO',
        done: false
    }

    const onToggleTodo = jest.fn();
    const onDeleteTodo = jest.fn();

    beforeEach(() => jest.clearAllMocks());


    test('se debe renderizar todoitem', () => {


        const { container } = render(<TodoItem
            todo={todo}
            onToggleTodo={onToggleTodo}
            onDelete={onDeleteTodo} />);



        const liElement = screen.getByRole('listitem');


        expect(container).toMatchSnapshot();
        expect(liElement.className).toBe('list-group-item d-flex justify-content-between');

        const spanElement = screen.getByLabelText('span');
        expect(spanElement.className.trim()).toBe('align-self-center')


    });

    test('Debe de mostrar el Todo Completado', () => {

        todo.done = true;

        render(<TodoItem
            todo={todo}
            onToggleTodo={onToggleTodo}
            onDelete={onDeleteTodo} />);




        const spanElement = screen.getByLabelText('span');
        expect(spanElement.className.trim()).toContain('text-decoration-line-through');


    });

    test('el Span debe llamar el ToggleTodo cuando se hace click', () => {


        render(<TodoItem
            todo={todo}
            onToggleTodo={ onToggleTodo }
            onDelete={ onDeleteTodo } />);

        const spanElement = screen.getByLabelText('span');
        fireEvent.click( spanElement );

        expect( onToggleTodo ).toHaveBeenCalledWith( todo.id )


    });

    test('el boton debe llamar el onDeleteTodo cuando se hace click', () => {


        render(<TodoItem
            todo={todo}
            onToggleTodo={ onToggleTodo }
            onDelete={ onDeleteTodo } />);

        const buttonElement = screen.getByRole('button');
        fireEvent.click( buttonElement );

        expect( onDeleteTodo ).toHaveBeenCalledWith( todo.id );


    });


});