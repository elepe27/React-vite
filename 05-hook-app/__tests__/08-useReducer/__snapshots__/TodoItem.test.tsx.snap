// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoItem/> se debe renderizar todoitem 1`] = `
<div>
  <li
    class="list-group-item d-flex justify-content-between"
  >
    <span
      aria-label="span"
      class="align-self-center "
    >
      TODO NUEVO
    </span>
    <button
      class="btn btn-sm btn-danger"
    >
       Borrar 
    </button>
  </li>
</div>
`;
